### PR TITLE
API Refactor

### PR DESCRIPTION
**Pull request by sbezverk**
This PR refactors NSM APIs to the level described in nsm_api.md

### CODE REVIEW

**Comment by sbezverk** [comment]
@edwarnicke @mestery @fkautz The gate is green and end to end test was restored.. Please review.

File: conf/sample/networkservice-daemonset.yaml [hunk 1]
```diff
@@ -14,6 +14,15 @@ spec:
         - name: netmesh
           image: networkservicemesh/netmesh
           imagePullPolicy: IfNotPresent
+          env:
+            - name: NSM_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: NSM_IPADDRESS
```
Comment (edwarnicke) [inline conf/sample/networkservice-daemonset.yaml]: Shouldn't the NSM_IPADDRESS be the PodIp for the NSM Pod?

File: docs/nsm_api.md [hunk 2]
```diff
@@ -359,16 +364,17 @@ NSE is the actual provider of a network service, to make aware NSM of the servic
...
 }
 
 message NetworkServiceEndpoint {
-   string network_service_name = 1;
-   string network_service_host = 2;
-   string nse_provider_name = 3;
-   string nse_provider_namesapce = 4;
-   string socket_location = 5;
-   repeated common.Interface interface = 6;
+    string network_service_name = 1;
```
Comment (edwarnicke) [inline docs/nsm_api.md]: We should have labels here as well (simple map<string,string> labels)

File: docs/nsm_api.md [hunk 3]
```diff
@@ -359,16 +364,17 @@ NSE is the actual provider of a network service, to make aware NSM of the servic
...
+    string network_service_host = 2;
+    string nse_provider_name = 3;
+    string nse_provider_namespace = 4;
+    string socket_location = 5;
```
Comment (edwarnicke) [inline docs/nsm_api.md]: Question... is the socket_location relative to the directory we mount in for the nsm resource?

I'm  trying to wrap my head around how we deal with the fact that a filesystem location in the NSE and the filesystem in the NSM are not the same file system, and so any location should be relative to some subdir mounted into both.

**Comment by edwarnicke** [comment]
Could we get comments explaining what the fields are for messages in the nsm_api.md for the new/changed fields in messages?

**Comment by edwarnicke** [comment]
It looks like this is also doing an upgrade of dependencies in addition to API refactor work.  This makes it a bit more combersome to review.  I do realize sometimes this kind of thing is unavoidable, but in the future, if possible, it would be good to break out such dependency upgrades into separate PRs :)

File: docs/nsm_api.md (hunk 2 above)
Comment (sbezverk) [reply docs/nsm_api.md]: I can add it, but could you clarify its purpose so it can be named accordingly.

**Comment by edwarnicke** [comment]
@sbezverk I don't feel strongly enough about the vendor changes to make you do work this time :)  Consider it a friendly request for the future.  I figured it was driven by something like your cluster running 1.11.13 and needing to do a lot of debugging, as you said.  

I'm a strong believer in good guidelines and good practice.  I also strongly believe there are times and places to diverge from them.  So don't read my statement *too* strongly :)

**Comment by edwarnicke** [comment]
@sbezverk I'm looking through your most recent patch adding comments.  Thank you for doing so, it helps.  We should probably decide at some point whether comments in the inline proto or comments *around* inline proto are more helpful to readers.  I've so far been using comments *in* the inline proto.  Please note I'm not asking you to change here, just opening the question :)

Also, many thanks for updating the api doc as you go, its super super good practice.

**Pull request merged**